### PR TITLE
Allow to specify subject (for CSRs) and issuer (for CRLs) ordered

### LINKED ISSUE #291: Problems with CRL issuer matching Cert issuer.
##### SUMMARY

Problems creating CRL with complex Issuer.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

community.crypto.x509_crl

##### ANSIBLE VERSION

```
ansible 2.9.1
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Nov  6 2016, 00:28:07) [GCC 4.8.5 20150623 (Red Hat 4.8.5-11)]

```

##### COLLECTION VERSION

```paste below
v1.9.3

```

##### CONFIGURATION

```paste below
[root@noc1 client-certs]# ansible-config dump --only-changed

CACHE_PLUGIN(/etc/ansible/ansible.cfg) = jsonfile
CACHE_PLUGIN_CONNECTION(/etc/ansible/ansible.cfg) = /var/lib/ansible-fact-cache/
CACHE_PLUGIN_TIMEOUT(/etc/ansible/ansible.cfg) = 3600
DEFAULT_GATHERING(/etc/ansible/ansible.cfg) = smart
[root@noc1 client-certs]# 

```

##### OS / ENVIRONMENT

Centos7

##### STEPS TO REPRODUCE

Create a complex issuer with CN and C and other values.

```
Creating a CRL with:
  - name: revoke the users
    community.crypto.x509_crl:
      path: "{{ cadir }}/crl.pem"
      privatekey_path: "{{ cadir }}/private/crlcacert.pem"
      mode: generate
      backup: true
      issuer:
        CN: testvm.sycle.net
      last_update: "+0s"
      next_update: "+7d"
      revoked_certificates:
        - path: "{{ cadir }}/users/alvin.pem"
       
Works but using:

 - name: revoke the users
    community.crypto.x509_crl:
      path: "{{ cadir }}/crl.pem"
      privatekey_path: "{{ cadir }}/private/crlcacert.pem"
      mode: generate
      backup: true
      issuer:
        C: CA
        ST: Ontario
        L: Toronto
        O: Sycle Inc.
        CN: testvm.sycle.net
      last_update: "+0s"
      next_update: "+7d"
      revoked_certificates:
        - path: "{{ cadir }}/users/alvin.pem"
        
does not
The self signed CA has an issuer of /C=CA/ST=Ontario/L=Toronto/O=Sycle Inc./CN=testvm.sycle.net
but the CRL has an issuer of /CN=testvm.sycle.net/C=CA/L=Toronto/O=Sycle Inc./ST=Ontario

makeing the issuer just the CN works.

```

##### EXPECTED RESULTS

The CRL check of a user cert should work

##### ACTUAL RESULTS

I get:
error 3 at 0 depth lookup:unable to get certificate CRL

```paste below
With a complex Issuer that is mis-ordered I get
openssl verify -crl_check -CRLfile crl.pem /etc/pki/CA/users/amir.pem 
/etc/pki/CA/users/amir.pem: C = CA, ST = Ontario, L = Toronto, O = Sycle Inc., CN = amir@sycle.net
error 3 at 0 depth lookup:unable to get certificate CRL


```

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
Fixes #291.

##### ISSUE TYPE
- Bugfix Pull Request
- Feature Pull Request

##### COMPONENT NAME
acme_certificate
openssl_csr
openssl_csr_pipe
x509_crl

### CODE REVIEW

**Comment by felixfontein** [comment]
ready_for_review

File: changelogs/fragments/315-ordered-names.yml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+minor_changes:
+  - "acme_certificate - the ``subject`` and ``issuer`` fields in in the ``select_chain`` entries are now stricter validated (https://github.com/ansible-collections/community.crypto/pull/316)."
```
Comment (Ajpantuso) [inline changelogs/fragments/315-ordered-names.yml:2]:
```suggestion
  - "acme_certificate - the ``subject`` and ``issuer`` fields in in the ``select_chain`` entries are now more strictly validated (https://github.com/ansible-collections/community.crypto/pull/316)."
```

File: plugins/module_utils/crypto/support.py [hunk 2]
```diff
@@ -237,16 +237,43 @@ def load_certificate_request(path, content=None, backend='cryptography'):
...
+                    raise ValueError('Values for {key} must be strings'.format(key=key))
                 result.append((key, entry))
+        elif isinstance(value, six.string_types):
+            result.append((key, value))
         else:
-            result.append((key, input_dict[key]))
+            if name_field_name:
+                raise ValueError(
+                    'Value for {key} in {name} must be either a string or a list of strings'.format(key=key, name=name_field_name))
+            raise ValueError('Value for {key} must be either a string or a list of strings'.format(key=key))
```
Comment (Ajpantuso) [inline plugins/module_utils/crypto/support.py:258]:
```suggestion
                raise TypeError(
                    'Value for {key} in {name} must be either a string or a list of strings'.format(key=key, name=name_field_name))
            raise TypeError('Value for {key} must be either a string or a list of strings'.format(key=key))
```

File: plugins/module_utils/crypto/support.py [hunk 3]
```diff
@@ -237,16 +237,43 @@ def load_certificate_request(path, content=None, backend='cryptography'):
...
+
+    result = []
+    for index, entry in enumerate(input_list):
+        if len(entry) != 1:
+            raise ValueError(
+                'Entry #{index} in {name} must be a dictionary with exactly one key-value pair'.format(
+                    name=name_field_name, index=index + 1))
+        try:
+            result.extend(parse_name_field(entry, name_field_name=name_field_name))
+        except ValueError as exc:
```
Comment (Ajpantuso) [inline plugins/module_utils/crypto/support.py:273]:
```suggestion
        except (TypeError, ValueError) as exc:
```

File: plugins/modules/x509_crl.py [hunk 4]
```diff
@@ -462,8 +476,16 @@ def __init__(self, module):
...
-        self.issuer = [(entry[0], entry[1]) for entry in self.issuer if entry[1]]
+        try:
+            if module.params['issuer_ordered']:
+                self.issuer_ordered = True
+                self.issuer = parse_ordered_name_field(module.params['issuer_ordered'], 'issuer_ordered')
+            else:
+                self.issuer_ordered = False
+                self.issuer = parse_name_field(module.params['issuer'], 'issuer')
+                self.issuer = [(entry[0], entry[1]) for entry in self.issuer if entry[1]]
+        except ValueError as exc:
```
Comment (Ajpantuso) [inline plugins/modules/x509_crl.py:487]:
```suggestion
        except (TypeError, ValueError) as exc:
```

**Review by Ajpantuso — CHANGES_REQUESTED** [review]
A few minor comments, but otherwise looks good.

File: plugins/module_utils/crypto/support.py [hunk 5]
```diff
@@ -237,16 +237,43 @@ def load_certificate_request(path, content=None, backend='cryptography'):
...
-    for key in input_dict:
-        if isinstance(input_dict[key], list):
-            for entry in input_dict[key]:
+    for key, value in input_dict.items():
+        if isinstance(value, list):
+            for entry in value:
+                if not isinstance(entry, six.string_types):
+                    if name_field_name:
+                        raise ValueError('Values for {key} in {name} must be strings'.format(key=key, name=name_field_name))
+                    raise ValueError('Values for {key} must be strings'.format(key=key))
```
Comment (felixfontein) [inline plugins/module_utils/crypto/support.py:250]:
```suggestion
                        raise TypeError('Values for {key} in {name} must be strings'.format(key=key, name=name_field_name))
                    raise TypeError(Values for {key} must be strings'.format(key=key))
```

**Comment by felixfontein** [comment]
@Ajpantuso what do you think of d497231e1cd2777ee7fe6bea856975d7b905101e and 196c85ffe25a171bd2b5f8fe27d38ea41eaf04bb? I think it's better to reject such values outright instead of silently ignoring them. The silent ignore makes sense for the explicit option fields for the CSR modules (or well, ignoring only `None` would make even more sense...), but I don't think it should happen for the other fields. What do you think?

**Comment by Ajpantuso** [comment]
> @Ajpantuso what do you think of [d497231](https://github.com/ansible-collections/community.crypto/commit/d497231e1cd2777ee7fe6bea856975d7b905101e) and [196c85f](https://github.com/ansible-collections/community.crypto/commit/196c85ffe25a171bd2b5f8fe27d38ea41eaf04bb)? I think it's better to reject such values outright instead of silently ignoring them. The silent ignore makes sense for the explicit option fields for the CSR modules (or well, ignoring only `None` would make even more sense...), but I don't think it should happen for the other fields. What do you think?

Makes sense to me. Silently ignoring could even lead users to believe that defaults would be applied from say their OpenSSL config. More explicit is better in this case.

**Review by Ajpantuso — APPROVED** [review]
LGTM

**Comment by felixfontein** [comment]
@Ajpantuso thanks again for reviewing this!

**Pull request merged**